File: site-presentation/src/lib/dce/extract.ts
// Pipeline steps 3-4 — text extraction with page anchoring (server-only).
//
// - PDF: native text via pdfjs-dist, one logical page per PDF page. Pages that
//   come back near-empty are treated as scanned images: the page is rasterized
//   (pdfjs + @napi-rs/canvas) and read by the autonomous Tesseract OCR engine
//   (see ocr.ts). A failed OCR never sinks the rest of the DCE — the page is
//   kept and flagged (CDC §6).
// - Image files (jpg/png/tiff/webp/bmp): OCR'd directly, single logical page.
// - DOCX: mammoth (raw text), single logical page (DOCX has no page model).
// - XLSX: SheetJS, one logical page per worksheet.

import type { DceSettings } from "./config";
import { ocrImage } from "./ocr";

export type PageText = { pageNumber: number; text: string; ocrUsed: boolean };
export type ExtractResult = { pages: PageText[]; warnings: string[] };

// Image file extensions handled directly by OCR (whole file = one page).
export const IMAGE_EXTS = [".jpg", ".jpeg", ".png", ".tif", ".tiff", ".webp", ".bmp"];

async function extractPdf(
  content: Buffer,
  filename: string,
  settings: DceSettings,
): Promise<ExtractResult> {
  const result: ExtractResult = { pages: [], warnings: [] };
  let pdfjs: typeof import("pdfjs-dist/legacy/build/pdf.mjs");
  try {
    pdfjs = await import("pdfjs-dist/legacy/build/pdf.mjs");
  } catch (err) {
    result.warnings.push(`${filename}: moteur PDF indisponible (${String(err)})`);
    return result;
  }

  const loadingTask = pdfjs.getDocument({ data: new Uint8Array(content), useSystemFonts: true });
  try {
    const doc = await loadingTask.promise;
    let ocrCount = 0;
    for (let i = 1; i <= doc.numPages; i++) {
      const page = await doc.getPage(i);
      const tc = await page.getTextContent();
      const parts: string[] = [];
      for (const item of tc.items) {
        if ("str" in item) {
          parts.push(item.str);
          if (item.hasEOL) parts.push("\n");
        }
      }
      let text = parts.join("").replace(/[ \t]+\n/g, "\n").trim();
      let ocrUsed = false;

      if (text.length < settings.ocrMinChars) {
        // Sparse page -> likely scanned. Rasterize it and OCR the image.
        if (settings.ocrEnabled && ocrCount < settings.ocrMaxPages) {
          try {
            const { createCanvas } = await import("@napi-rs/canvas");
            const viewport = page.getViewport({ scale: settings.ocrScale });
            const canvas = createCanvas(Math.ceil(viewport.width), Math.ceil(viewport.height));
            const ctx = canvas.getContext("2d");
            await page.render({
              canvasContext: ctx as unknown as CanvasRenderingContext2D,
              canvas: canvas as unknown as HTMLCanvasElement,
              viewport,
            }).promise;
            const ocrText = await ocrImage(canvas.toBuffer("image/png"));
            ocrCount++;
            if (ocrText.length > text.length) {
              text = ocrText;
              ocrUsed = true;
            }
            if (!ocrText) {
              result.warnings.push(`${filename} p.${i}: page scannée — OCR sans texte lisible.`);
            }
          } catch (err) {
            result.warnings.push(`${filename} p.${i}: OCR échoué (${String(err)}).`);
          }
        } else {
          const why = settings.ocrEnabled ? "quota OCR du document atteint" : "OCR désactivé";
          result.warnings.push(
            `${filename} p.${i}: peu ou pas de texte — page probablement scannée (${why}).`,
          );
        }
      }

      result.pages.push({ pageNumber: i, text, ocrUsed });
      page.cleanup();
    }
  } catch (err) {
    result.warnings.push(`${filename}: PDF illisible (${String(err)})`);
  } finally {
    await loadingTask.destroy().catch(() => {});
  }
  return result;
}

async function extractDocx(content: Buffer, filename: string): Promise<ExtractResult> {
  const result: ExtractResult = { pages: [], warnings: [] };
  try {
    const mammoth = (await import("mammoth")).default;
    const { value } = await mammoth.extractRawText({ buffer: content });
    result.pages.push({ pageNumber: 1, text: (value ?? "").trim(), ocrUsed: false });
  } catch (err) {
    result.warnings.push(`${filename}: DOCX illisible (${String(err)})`);
  }
  return result;
}

async function extractXlsx(content: Buffer, filename: string): Promise<ExtractResult> {
  const result: ExtractResult = { pages: [], warnings: [] };
  try {
    const XLSX = await import("xlsx");
    const wb = XLSX.read(content, { type: "buffer" });
    wb.SheetNames.forEach((name, index) => {
      const ws = wb.Sheets[name];
      const rows = XLSX.utils.sheet_to_json<unknown[]>(ws, { header: 1, blankrows: false });
      const lines = [`[Feuille: ${name}]`];
      for (const row of rows) {
        const cells = (row as unknown[]).filter((v) => v != null && v !== "").map((v) => String(v));
        if (cells.length) lines.push(cells.join(" | "));
      }
      result.pages.push({ pageNumber: index + 1, text: lines.join("\n"), ocrUsed: false });
    });
  } catch (err) {
    result.warnings.push(`${filename}: XLSX illisible (${String(err)})`);
  }
  return result;
}

async function extractImage(
  content: Buffer,
  filename: string,
  settings: DceSettings,
): Promise<ExtractResult> {
  const result: ExtractResult = { pages: [], warnings: [] };
  if (!settings.ocrEnabled) {
    result.pages.push({ pageNumber: 1, text: "", ocrUsed: false });
    result.warnings.push(`${filename}: image non lue (OCR désactivé).`);
    return result;
  }
  try {
    const text = await ocrImage(content);
    result.pages.push({ pageNumber: 1, text, ocrUsed: true });
    if (!text) {
      result.warnings.push(`${filename}: image sans texte lisible (OCR vide).`);
    }
  } catch (err) {
    result.pages.push({ pageNumber: 1, text: "", ocrUsed: false });
    result.warnings.push(`${filename}: OCR de l'image échoué (${String(err)}).`);
  }
  return result;
}

export async function extractText(
  filename: string,
  content: Buffer,
  settings: DceSettings,
): Promise<ExtractResult> {
  const lower = filename.toLowerCase();
  if (lower.endsWith(".pdf")) return extractPdf(content, filename, settings);
  if (lower.endsWith(".docx")) return extractDocx(content, filename);
  if (lower.endsWith(".xlsx")) return extractXlsx(content, filename);
  if (IMAGE_EXTS.some((ext) => lower.endsWith(ext))) {
    return extractImage(content, filename, settings);
  }
  return { pages: [], warnings: [`${filename}: format non supporté`] };
}

File: backend/app/pipeline/extract.py
"""Pipeline steps 3-4 — text extraction with page anchoring + OCR fallback.

- PDF: native text via PyMuPDF; pages that come back near-empty are treated as
  scanned images and re-read via Tesseract (``fra``).
- DOCX: python-docx (paragraphs + tables), single logical page.
- XLSX: openpyxl, one logical page per worksheet.

Robustness (CDC §6): a failed OCR on one page never sinks the rest of the DCE.
The page is kept with whatever text we have and flagged in ``warnings``.
"""

from __future__ import annotations

import io
import logging
from dataclasses import dataclass, field

from app.config import Settings

log = logging.getLogger(__name__)


@dataclass
class PageText:
    page_number: int  # 1-indexed
    text: str
    ocr_used: bool = False


@dataclass
class ExtractResult:
    pages: list[PageText] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


def _ocr_page(page, settings: Settings) -> str:
    """Render a PDF page to an image and OCR it. Returns "" on any failure."""
    try:
        import pytesseract
        from PIL import Image

        pix = page.get_pixmap(dpi=settings.ocr_dpi)
        img = Image.open(io.BytesIO(pix.tobytes("png")))
        return pytesseract.image_to_string(img, lang=settings.ocr_lang) or ""
    except Exception:  # noqa: BLE001 — OCR is best-effort
        log.exception("OCR failed on page %s", getattr(page, "number", "?"))
        return ""


def extract_pdf(content: bytes, filename: str, settings: Settings) -> ExtractResult:
    import fitz  # PyMuPDF

    result = ExtractResult()
    try:
        doc = fitz.open(stream=content, filetype="pdf")
    except Exception as exc:  # noqa: BLE001
        result.warnings.append(f"{filename}: PDF illisible ({exc})")
        return result

    with doc:
        for index in range(doc.page_count):
            page = doc.load_page(index)
            page_no = index + 1
            native = (page.get_text("text") or "").strip()
            if len(native) >= settings.ocr_min_chars:
                result.pages.append(PageText(page_no, native, ocr_used=False))
                continue
            # Sparse page -> likely scanned. Try OCR, keep the better of the two.
            ocr = _ocr_page(page, settings).strip()
            if len(ocr) > len(native):
                result.pages.append(PageText(page_no, ocr, ocr_used=True))
                if not ocr:
                    result.warnings.append(f"{filename} p.{page_no}: page illisible (OCR vide)")
            else:
                result.pages.append(PageText(page_no, native, ocr_used=False))
                if not native:
                    result.warnings.append(f"{filename} p.{page_no}: page illisible")
    return result


def extract_docx(content: bytes, filename: str) -> ExtractResult:
    import docx  # python-docx

    result = ExtractResult()
    try:
        document = docx.Document(io.BytesIO(content))
    except Exception as exc:  # noqa: BLE001
        result.warnings.append(f"{filename}: DOCX illisible ({exc})")
        return result

    parts: list[str] = [p.text for p in document.paragraphs if p.text.strip()]
    for table in document.tables:
        for row in table.rows:
            cells = [c.text.strip() for c in row.cells]
            if any(cells):
                parts.append(" | ".join(cells))
    # DOCX has no reliable page model; treat as a single logical page.
    result.pages.append(PageText(1, "\n".join(parts), ocr_used=False))
    return result


def extract_xlsx(content: bytes, filename: str) -> ExtractResult:
    import openpyxl

    result = ExtractResult()
    try:
        wb = openpyxl.load_workbook(io.BytesIO(content), read_only=True, data_only=True)
    except Exception as exc:  # noqa: BLE001
        result.warnings.append(f"{filename}: XLSX illisible ({exc})")
        return result

    for index, ws in enumerate(wb.worksheets, start=1):
        lines: list[str] = [f"[Feuille: {ws.title}]"]
        for row in ws.iter_rows(values_only=True):
            cells = [str(v) for v in row if v is not None]
            if cells:
                lines.append(" | ".join(cells))
        result.pages.append(PageText(index, "\n".join(lines), ocr_used=False))
    wb.close()
    return result


def extract_text(filename: str, content: bytes, settings: Settings) -> ExtractResult:
    """Dispatch on extension. Unknown types return an empty result + a warning."""
    lower = filename.lower()
    if lower.endswith(".pdf"):
        return extract_pdf(content, filename, settings)
    if lower.endswith(".docx"):
        return extract_docx(content, filename)
    if lower.endswith(".xlsx"):
        return extract_xlsx(content, filename)
    return ExtractResult(warnings=[f"{filename}: format non supporté"])
